Use `UTxO era` as single source of truth in balanceTransaction (second take)

Anviking opened the pull request:
- [x] Ensure all of `balanceTx` sees a consistent `UTxO` set / input resolution.
    - [x] Let `assignScriptRedeemers` take `Ledger.UTxO era` easily converted from `Cardano.UTxO era`
    - [x] Create a `combinedUTxO :: Cardano.UTxO era` once in balanceTx as single source of truth for all input lookup
    - **New error**: `unresolved_inputs ` replacing undefined behaviour
    -  **New error**: `input_resolution_conflicts` replacing undefined behaviour

### Comments

- This PR is like #3255 but only with minimal changes to `assignScriptRedeemers` and no changes to `NetworkLayer`.
- Replacing `Cardano.UTxO` with `Ledger.UTxO` in the interface to `balanceTransaction` can be done later.

### Issue Number

ADP-1662

Anviking committed:
Add more conversion functions

Anviking committed:
Use a `UTxO era` as single source of truth in balanceTx

Core goal: replace ad-hoc output/utxo type with `Cardano.UTxO` here

```
 data PartialTx era = PartialTx
     { tx :: Cardano.Tx era
-    , inputs :: [(TxIn, TxOut, Maybe (Hash "Datum"))]
+    , inputs :: Cardano.UTxO era
```

How?

- Convert from API types to `Cardano.UTxO` once in the server handler
- Bonus: add `guardWalletUTxOConsistencyWith inputUTxO`
  - Need is verified by property test.

This results in two new errors:
- unresolved_inputs
- input_resolution_conflicts
each replacing undefined behaviour.

NOTE: `Cardano.UTxO era` is isomorphic to `Ledger.UTxO` era, so this
commit /is/ a step forward w.r.t. moving towards ledger types (as
opposed to wallet and cardano-api types).

Includes squashed review suggestions from Jonathan Knowles.

Anviking commented on lib/core/src/Cardano/Api/Extra.hs:
```diff
@@ -56,3 +64,9 @@ asAnyShelleyBasedEra = \case
+deriving instance Eq ScriptExecutionError
+
+-- NOTE: Needs Eq PastHorizonException orphan
+deriving instance Eq TransactionValidityError
```
I don't think this is needed anymore (everything in the module)

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -694,42 +697,56 @@ _evaluateTransactionBalance
+        --
+        -- If the overridden outputs are incorrect, the wallet will
+        -- incorrectly calculate the balance, and the transaction
+        -- will ultimately be rejected by the node.
```
Suggestion: perhaps move some or all of this comment to the Haddock comment for `_toCardanoUTxO`, or alternatively the interface function `toCardanoUTxO`.

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -694,42 +697,56 @@ _evaluateTransactionBalance
+        -- incorrectly calculate the balance, and the transaction
+        -- will ultimately be rejected by the node.
+        --
+        -- If the overwridden outputs simply adds datum hashes
```
```suggestion
        -- If the overwritten outputs simply add datum hashes
```

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -694,42 +697,56 @@ _evaluateTransactionBalance
+    => UTxO
+    -> [(TxIn, TxOut, Maybe (Hash "Datum"))]
+    -> Cardano.UTxO era
+_toCardanoUTxO utxo extraUTxO =
```
Suggestion: find a way to combine this function and `Shelley.Compatibility.toCardanoUTxO`.

Perhaps it would be helpful to use the following signature:
```hs
toCardanoUTxO
    :: ShelleyBasedEra era
    -> Map TxIn (TxOut, Maybe (Hash "Datum"))
    -> Cardano.UTxO era
```

So within `balanceTransaction`, we'd have something like:

```
toCardanoUTxO (Map.fromList extraUTxO <> fmap (\o -> (o, Nothing)) UTxO.toMap)
```

Anviking replied:
This is a very good suggestion, but I'm leaning towards not touching this anymore today and either 1) opening a second pr or 2) exploring a more proper a solution in #3436 when I get back by replacing `[(TxIn, TxOut, Maybe (Hash "Datum"))]` with `Ledger.UTxO` and having the conversions in `core` already. I'd like this latter actually.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3594,7 +3592,7 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx') seed
         let combinedUTxO = mconcat
                 [ resolvedInputsUTxO Cardano.ShelleyBasedEraAlonzo partialTx
-                , toCardanoUTxO Cardano.ShelleyBasedEraAlonzo walletUTxO
+                , Compatibility.toCardanoUTxO Cardano.ShelleyBasedEraAlonzo walletUTxO
```
Suggestion: here we use the compatibility function rather than the similarly-named function in `TransactionLayer`, but if these were merged into the same function (see comment above), then that single function would be tested by this test and also the tests for `balanceTransaction`.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1567,19 +1567,19 @@ normalizeDelegationAddress s addr = do
 data PartialTx era = PartialTx
     { tx :: Cardano.Tx era
-    , inputs :: [(TxIn, TxOut, Maybe (Hash "Datum"))]
+    , inputs :: Cardano.UTxO era
```
Add note about naming of the field

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1803,6 +1779,36 @@ balanceTransactionWithSelectionStrategy
+    --
+    -- Requires @guardWalletUTxOConsistencyWith inputUTxO@ to validate
+    -- for balancing to succeed.
+    extractExternallySelectedUTxO
```
Suggestion: could you either:

- add some examples of how this function is supposed to work to the comment?
or
- add some unit tests that demonstrate how the function works
or
- add some simple properties to document how the function is intended to work

Anviking replied:
done

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -3588,6 +3626,8 @@ data ErrBalanceTx
     | ErrBalanceTxAssignRedeemers ErrAssignRedeemers
     | ErrBalanceTxInternalError ErrBalanceTxInternalError
     | ErrBalanceTxZeroAdaOutput
+    | ErrBalanceTxConflictingResolvedInputs
```
Suggestion:

The following name captures the meaning of a conflict that occurs during the _process_ of input resolution:

```suggestion
     | ErrBalanceTxInputResolutionConflict
```

Anviking replied:
I added a `NonEmpty conflicts` argument and went with `ErrBalanceTxInputResolutionConflicts`, but I realised `conflicts` now can be read as a verb 🤔 The ambiguity seems a bit annoying, but going with it for now.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -3588,6 +3626,8 @@ data ErrBalanceTx
     | ErrBalanceTxInternalError ErrBalanceTxInternalError
     | ErrBalanceTxZeroAdaOutput
+    | ErrBalanceTxConflictingResolvedInputs
+    | ErrBalanceTxUnresolvedInputs [TxIn]
```
Suggestion: use `NonEmpty`:
```suggestion
    | ErrBalanceTxUnresolvedInputs (NonEmpty TxIn)
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4862,7 +4864,17 @@ instance IsServerError ErrBalanceTx where
+            apiError err403 UnresolvedInputs $ mconcat
+                [ "There are inputs in the transaction which I cannot find the "
+                , "corresponding outputs for."
+                ]
```
```suggestion
            apiError err403 UnresolvedInputs $ T.unwords
                [ "There are inputs in the transaction for which corresponding outputs"
                , "could not be found."
                ]
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4862,7 +4864,17 @@ instance IsServerError ErrBalanceTx where
+            apiError err403 ConflictingResolvedInputs $ mconcat
+                [ "At least one of the inputs you told me about had different "
+                , "ada quantity, assets, or address compared to the wallet UTxO"
+                , "Please ensure the resolved inputs are correct."
```
```suggestion
             apiError err403 ConflictingResolvedInputs $ T.unwords
                [ "At least one of the inputs you've told me about has an"
                , "asset quantity or address that is different from that"
                , "recorded in the wallet's UTxO set."
                ]
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -5344,12 +5356,12 @@ instance IsServerError ErrUpdateSealedTx where
-                , "redeemers:", pretty r <> ";"
-                , "Its execution is failing with the following error:"
-                , T.pack failure <> "."
+                , "redeemers." <> T.pack (show r) -- FIXME revert/polish
```
Suggestion: revert if you think this is no longer necessary.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -5392,6 +5404,11 @@ instance IsServerError ErrAssignRedeemers where
                 ]
+        ErrAssignRedeemersUnexpectedCardanoAPIError e ->
+            apiError err500 TranslationError $ T.unwords
+                [ "I got an error I didn't expect:"
```
Suggestion: remove this completely, as we no longer use the Cardano API (unless I'm mistaken).

jonathanknowles commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -1824,6 +1824,7 @@ data ApiErrorCode
     | UnableToDetermineCurrentEpoch
     | UnexpectedError
     | UnresolvedInputs
+    | ConflictingResolvedInputs
```
```suggestion
     | InputResolutionConflict
```

jonathanknowles commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -524,17 +523,12 @@ data ErrMkTransaction
     -- ^ Redeemer's data isn't a valid Plutus' data.
     | ErrAssignRedeemersTranslationError (TranslationError StandardCrypto)
-    -- ^ Mistranslating of hashes, credentials, certificates etc.
+    | ErrAssignRedeemersUnexpectedCardanoAPIError String
```
Suggestion: remove this, as it's no longer needed.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3163,75 +3162,132 @@ instance Buildable a => Show (ShowBuildable a) where
+          , val' <- prependOriginal shrink val
+          , dat' <- prependOriginal shrink dat
+          , refScript' <- prependOriginal (const []) refScript
+          ]
```
Suggestion: if you like (and have time), consider using `genShrinkSequence` to determine that shrinking terminates.

Anviking replied:
I did some "manual mutation testing" — I removed `guardWalletConsistencyWithUTxO` check — and saw the failure's shrinking at least terminate... to a different error message... I'd like to have a look at these kinds of things at some point in the future, but skipping for now.

Anviking replied:
> at least terminate... to a different error message.

Might have been because I switched to `shrinkList` for the `extraUTxO` which I now reverted

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3163,75 +3162,132 @@ instance Buildable a => Show (ShowBuildable a) where
+          map
+              (Cardano.TxOutValue Cardano.MultiAssetInAlonzoEra
+                  . toCardanoValue)
+              (shrink $ Compatibility.fromCardanoValue val)
```
Suggestion: it might be shorter to use `shrinkMapBy` here.

Anviking replied:
I suspect the GADTs will mess it up here in the case expression, so will refrain from attempting. When switch to using Cardano.Ledger.Mary.Value it would probably work out.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3163,75 +3162,132 @@ instance Buildable a => Show (ShowBuildable a) where
+        where
+          safeFromCardanoLovelace = fromCardanoLovelace . abs
+
 instance Arbitrary (PartialTx Cardano.AlonzoEra) where
```
Question: does this generator ever produce values of `PartialTx` that always cause `balanceTransaction` to fail with an error?

Anviking replied:
Yes, at least the following errors:
```
      15.9% output below minCoinValue
      11.5% conflicting networks
       9.7% not yet supported: zero ada output
       0.8% existing collateral
```
which are only related to the `PartialTx` and not the `Wallet'` or combination of the two. ("output below minCoinValue" would be specific to the PParams though)

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3163,75 +3162,132 @@ instance Buildable a => Show (ShowBuildable a) where
+        ] <>
+        [ restrictResolution $ PartialTx tx' inputUTxO redeemers
+        | tx' <- shrinkTxBabbage tx
+        ]
```
Question: is there any opportunity to reduce duplication between Alonzo and Babbage instances?

Anviking replied:
I tried making the balanceTx validity property use `BabbageEra` instead to be able to remove the `Alonzo` arbitrary instance, thinking while we'd ideally test both, this would be better. However, I ran into problems with keeping generators good enough for all `prop_balanceTransactionExistingReturnCollateral`,  `prop_balanceTransactionExistingTotalCollateral` and `prop_balanceTransactionValid`. 

I'd like to deal with this more properly so I'm leaving things as they are (with duplication) for now. (Btw, I'd like to remove the collateral properties in favour for a new check in `prop_balanceTransactionValid`.)

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3163,75 +3162,132 @@ instance Buildable a => Show (ShowBuildable a) where
+    => Cardano.UTxO era
+    -> [Cardano.UTxO era]
+shrinkInputResolution =
+     map utxoFromList . shrinkUTxOEntries . utxoToList
```
Suggestion: use `shrinkMapBy` if possible.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3163,75 +3162,132 @@ instance Buildable a => Show (ShowBuildable a) where
+    -- Recurse to shrink subsequent elements on their own
+    , map ((i,o):) (shrinkUTxOEntries rest)
+    ]
+shrinkUTxOEntries [] = []
```
Suggestion: use some combination of `shrinkMapBy` and `shrinkList`.

jonathanknowles commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3735,21 +3792,6 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx') seed
-        ins' = flip map ins $ \(i,o,dh) ->
-            case UTxO.lookup i walletUTxO of
-                Just o' -> (i,o',Nothing)
-                Nothing -> (i,o,dh)
```
👍🏻

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -5077,6 +5077,19 @@ x-errExistingKeyWitnesses: &errExistingKeyWitnesses
       type: string
       enum: ['existing_key_witnesses']
 
+x-errConflictingInputResolution: &errConflictingInputResolution
```
```suggestion
 x-errInputResolutionConflict: &errInputResolutionConflict
 ```
 
 (for consistency)

jonathanknowles commented on specifications/api/swagger.yaml:
```diff
@@ -5077,6 +5077,19 @@ x-errExistingKeyWitnesses: &errExistingKeyWitnesses
+      type: string
+      description: |
+        Occurs when providing an input resolution (input -> output mapping) as part of the
+        `inputs` field which conflicts with the wallet UTxO set when balancing.
```
Suggestion: copy what the `Api.Server` error description has?

Anviking replied:
Btw, I noticed the api docs for errors are kind of weird; the description looks specific to the `message` field rather than to the error, and I'm not sure to what extend the error code make sense:

<img width="826" alt="Skärmavbild 2022-08-30 kl  15 35 03" src="https://user-images.githubusercontent.com/304423/187451286-4015c65c-123a-4e16-b9f8-758441b97994.png">

Also `err400 bad request` seemed more appropriate than `err403 forbidden` for `unresolved_inputs` and `input_resolution_conflicts`, so I went with that.

jonathanknowles reviewed (approved):
This looks great to me. Just some suggestions (see comments).

Anviking committed:
Tweak doc comment for extractExternallySelectedUTxO

- Add examples
- I was unsure whether guardWalletUTxOConsistencyWith actually was
  needed for @extractExternallySelectedUTxO@ specifically, so I removed
  that statement. Tests should make sure @guardWalletUTxOConsistencyWith@
  is called anyway.

Anviking committed:
Indent and document shrinkUTxOEntries

Anviking committed:
Add note about PartialTx.inputs naming
